aiohttp/connector.py

```diff
@@ -170,6 +171,67 @@ def close(self) -> None:
         pass
 
 
+class KeyedCondition:
+    def __init__(self):
+        # NOTE: asyncio.Lock + asyncio.Condition are FIFO
+        self._keyed_cond: Dict[ConnectionKey, asyncio.Condition] = dict()
+
+        # if we wanted to peek into asyncio.Condition._waiters this variable would not be necessary
+        self._num_waiters: Dict[ConnectionKey, int] = dict()
+
+    def keys(self) -> KeysView[ConnectionKey]:
+        return self._keyed_cond.keys()
+
+    async def wait(self, key: ConnectionKey):
+        cond = self._keyed_cond.get(key)
+        if not cond:
+            cond = self._keyed_cond[key] = asyncio.Condition()
+
+        async with cond:
+            try:
+                if key not in self._num_waiters:
+                    self._num_waiters[key] = 1
+                else:
+                    self._num_waiters[key] += 1
+
+                await cond.wait()
+            finally:
+                self._num_waiters[key] -= 1
+
+    def _notify_key(self, cond: asyncio.Condition, key: ConnectionKey):
+        # if we made it here the condition may have gotten cleared so
+        # so we may still KeyError on the line below
+        assert self._num_waiters[key] > 0
+        cond.notify(1)
+
+        # it's critical that there are now async calls below this line to
+        # assure consistency of self._keyed_cond
+        self._num_waiters[key] -= 1
+
+        if self._num_waiters[key] == 0:
+            del self._num_waiters[key]
+            del self._keyed_cond[key]
+
+    async def notify_random(self):
+        for key in random.shuffle(self._keyed_cond.keys()):
```

Comment:
`random.shuffle` works in place
```suggestion
        keys = list(self.keys())
        random.shuffle(keys)
        for key in keys:
```

Reply:
thanks